### ISSUE TITLE
make coverage command which found some error to create html coverage files

### ISSUE THREAD

**Issue by alishan121312** [ZipCPU/axidmacheck#3]
`/axidmacheck/sim$ ./main_tb -d

Opening Bus-master with
	VCD File         = trace.vcd
AXIMM2S Check:
	BEATS:  0x00002000
	CLOCKS: 0x00002000
	COUNTS: 0x00002031
AXIMM2S (abort) Check:
	BEATS:  0x000002f7
	CLOCKS: 0x000002f7
	COUNTS: 0x00000328
AXIMM2S (unaligned) Check: No unaligned support (0x01000024)
AXIMM2S (continuous) Midway:
	BEATS:  0x00002000
	CLOCKS: 0x00002000
	COUNTS: 0x00002027
AXIMM2S (continuous) Midway:
	STATUS: 0x30090000
	BEATS:  0x00004000
	CLOCKS: 0x000041e7
	COUNTS: 0x00004218
AXIS2MM Check:
	COUNTS: 0x00002135
	ERR-CODE: 0
AXIS2MM (abort) Check:
	COUNTS: 0x00000206
	ERR-CODE: 0
Running AXI DMA test
Test has begin
AXIDMA Check:
%Error: %Error: Can't write 'logs/coverage.dat'
Aborting...
Aborted (core dumped)
`
Makefile not able to create log file and found the above error .kindly guide me for above issues.

thanks in advance

**Comment by ZipCPU** [comment]
Which branch are you using?

**Comment by ZipCPU** [comment]
NVM, I see the issue in the main branch.

For now, just create a "logs" subdirectory of the sim/ directory.  I'll have to come back and adjust either automaster_tb.cpp or the Makefile to make sure that file exists in the future.

Dan

**Comment by alishan121312** [comment]
Hi Dan
I resolved that issue and thanks for your quick reply. I need some assistance regarding the axidma MM2M register details. Are you able to share me any source which you explain the axidma MM2M (https://github.com/ZipCPU/wb2axip/blob/master/rtl/axidma.v ) code flow with register summary? 
Thanks you in advance 
regards

**Comment by ZipCPU** [comment]
So, the "document" I have to describe the operation of the AXIDMA is the axidma.v file itself.  Pay attention to the notes at the top of the file.  There are 8x 32-bit registers grouped so they can also be interpreted as one 32-bit register and 3x 64bit registers.  The first 32-bit register is the control/status register.  This is what kicks off the transfer.  You typically write to that one last.  The 3 64-bit registers are the source address, the destination address, and the length of the transfer (in bytes).   The upper 32-bits of each of these registers will be unused if the address bus has 32-bits or less.  Otherwise, these registers are ordered in little endian format, with the low address first and the high address last.  If OPT_UNALIGNED is set, then any length or address, aligned or not, are acceptable.

Let's come back to that control register.
1. To start a transfer, write a 1'1 to the register.
2. If you also write to bit [2], the DMA will generate an interrupt upon completion.
3. Once the DMA begins, bit[3] (busy) will turn high.  This is how you know the DMA is active.
4. Bit[4] is an AXI bus error bit.  If set, it means the DMA operation ended early due to a bus error.  This bit must be cleared before any subsequent transfer can begin.  You can clear this by writing a 1 to this bit any time the DMA is idle.  This bit may also be cleared at the same time the DMA is started by writing a one to it at that time.  Fr this reason, I'll often (automatically) write to both the ERR and the START bit at the same time.
5. The ABORT_KEY, currently set to 8'h6d, is a special value which, if written to the MSB of the control word, will cause any ongoing operation to abort.  If an operation has been aborted due to this type of user request, then bit[5] of the control word will be set.  It may be cleared by writing a 1 to the bit, by starting another operation, or by writing a 1 to the error bit.
6. Bits 19:16 control the AXI QOS field used by the DMA.  These bits are write only.  (Perhaps that's an oversight ...)
7. Bits 22:20 control the AXI PROT field used by the DMA.  These bits are also write only.  Whether this or the QOS field are useful, or security violations, is up to the user to determine based upon whether or not you wire these up or override them internally.  For now ... they are available to be set.

While I have some C code lying around for testing the DMA, it's not yet a part of the repository so I'm not sure how well I can vouch for it.  In general, though, the code for driving this might look like:

    void dma_memcpy(char *dst, const void *src, size_t n) { // Assuming 32-bit pointers ...
        _dma->src = src;
        _dma->dst = dst;
        _dma->len = len;
        _dma->control = 0x17; // Clear err, enable interrupts, clear int, and start DMA

        while(_dma->control & 8)
            ;
    }

Interrupts would just be a bit more complicated, but not by much.  In the case of an interrupt, you'd start it as above and then instead of the `while` loop, you'd wait for the interrupt to indicate that it was done.

Is that what you are looking for?

Dan

**Comment by alishan121312** [comment]
Hi Dan
Thanks for your kind response.I think that will be very helpful for me as you explained above.what about the other bit in control register?
other bits of control register like 6 to 18 and 23 to 31 are reserved ?

**Comment by ZipCPU** [comment]
Of course!

Reserved bits will read as zero.  (As do the r_prot and r_qos bits.)  Writes to other bits are also ignored.

Dan

**Comment by alishan-rs** [comment]
Hi Dan 
Is it possible to test the axisgdma by using that test bench  ?

**Comment by ZipCPU** [comment]
The SG DMA is not (currently) hooked up to the main RTL.  Until the two are connected, it will not be possible.  Specifically, the SG DMA will need to connect into the interconnect as a new master.  Connecting it would require generating an AutoFPGA configuration file for it, but otherwise would be fairly uneventful.  From there a test script would need to be generated, as I don't have one at present.

Bottom line is that it would take about a half days work to set it up.

Dan

**Comment by alishan121312** [comment]
Hi Dan
is it possible to to find the functional coverage of AXIDMACHECK?

**Comment by ZipCPU** [comment]
Yes/no.  I've got the coverage data on my desk.  However, it depends upon a file someone gave me and ... I'm not sure what the appropriate license is for that file, so I don't feel like I can release it.  So while I can calculate the coverage, I can't really post how I've done so until that issue is resolved.

**Comment by ZipCPU** [comment]
The three DMAs get:

MM2S: 91%
S2MM: 86%
AXIDMA: 89%

Better coverage would require 1) checking read returns for more than just the status bits, and 2) an AXI environment that would actually generate backpressure.  For the S2MM, you'd also want to check the overflow catch circuit on the input which the TB can't currently generate.

Dan

**Comment by alishan121312** [comment]
ok great.
Is it to do Post-synthesis Simulation of axidma?

**Comment by ZipCPU** [comment]
Are you asking for the purpose of this repository?

Think of this as more of a "post-formal" simulation of AXIDMA.  All of the designs within this repository have been formally verified.  The purpose of this repository is to sort of shake things out, and see if the formal proof missed anything.

**Comment by alishan121312** [comment]
I'm currently working with the AXI DMA and have been testing its functionality. I connected an AXI RAM to the master side of the DMA and started reading data from a specific source address, then writing it to a specific destination address. However, I have noticed that the data being read from the source address is not exactly what I was expecting, even though the control register was correctly configured. I also confirmed that the read address is the same as what I configured on the master side of the AXI DMA.
I was wondering if you could offer any guidance on what might be causing this issue. I have already double-checked my control register settings and confirmed that the source address is correct on master side of AXI DMA during ready address. However, the data being read from that address of AXI RAM is not what I was expecting.

**Comment by ZipCPU** [comment]
The classic problem is usually cached data.  I usually need to clear my data cache after any DMA operation.  If that's not it, then I'll need more information in order to help debug what's going on.

**Comment by ZipCPU** [comment]
I'm going to close this issue, since its no longer about `make coverage`.  If you have a problem with the AXI DMA, please open another issue and we'll continue the discussion there.
